File: frontend/appflowy_flutter/lib/plugins/trash/application/trash_service.dart
import 'dart:async';

import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-folder/trash.pb.dart';
import 'package:appflowy_result/appflowy_result.dart';

class TrashService {
  Future<FlowyResult<RepeatedTrashPB, FlowyError>> readTrash() {
    return FolderEventListTrashItems().send();
  }

  static Future<FlowyResult<void, FlowyError>> putback(String trashId) {
    final id = TrashIdPB.create()..id = trashId;

    return FolderEventRestoreTrashItem(id).send();
  }

  Future<FlowyResult<void, FlowyError>> deleteViews(List<String> trash) {
    final items = trash.map((trash) {
      return TrashIdPB.create()..id = trash;
    });

    final ids = RepeatedTrashIdPB(items: items);
    return FolderEventPermanentlyDeleteTrashItem(ids).send();
  }

  Future<FlowyResult<void, FlowyError>> restoreAll() {
    return FolderEventRecoverAllTrashItems().send();
  }

  Future<FlowyResult<void, FlowyError>> deleteAll() {
    return FolderEventPermanentlyDeleteAllTrashItem().send();
  }
}

File: frontend/appflowy_tauri/src/appflowy_app/application/folder/trash.service.ts
import {
  FolderEventListTrashItems,
  FolderEventPermanentlyDeleteAllTrashItem,
  FolderEventPermanentlyDeleteTrashItem,
  FolderEventRecoverAllTrashItems,
  FolderEventRestoreTrashItem,
  RepeatedTrashIdPB,
  TrashIdPB,
} from '@/services/backend/events/flowy-folder';

export const getTrash = async () => {
  const res = await FolderEventListTrashItems();

  if (res.ok) {
    return res.val.items;
  }

  return [];
};

export const putback = async (id: string) => {
  const payload = new TrashIdPB({
    id,
  });

  const res = await FolderEventRestoreTrashItem(payload);

  if (res.ok) {
    return res.val;
  }

  return Promise.reject(res.err);
};

export const deleteTrashItem = async (ids: string[]) => {
  const items = ids.map((id) => new TrashIdPB({ id }));
  const payload = new RepeatedTrashIdPB({
    items,
  });

  const res = await FolderEventPermanentlyDeleteTrashItem(payload);

  if (res.ok) {
    return res.val;
  }

  return Promise.reject(res.err);
};

export const deleteAll = async () => {
  const res = await FolderEventPermanentlyDeleteAllTrashItem();

  if (res.ok) {
    return res.val;
  }

  return Promise.reject(res.err);
};

export const restoreAll = async () => {
  const res = await FolderEventRecoverAllTrashItems();

  if (res.ok) {
    return res.val;
  }

  return Promise.reject(res.err);
};
